### ISSUE TITLE
Application's crash

### ISSUE THREAD

**Issue by garakh** [01org/cordova-plugin-intel-xdk-device#1]
We've got our apps been crashed today. Looks like this commit https://github.com/01org/cordova-plugin-intel-xdk-device/commit/840f22ff7f0d7d490fb45b354c854675740a10b9 has broken something up.

We use this plugin in Intel XDK with cordova and it worked fine few days ago.

**Comment by tony--** [comment]
@garakh - Can you tell me how to reproduce the issue?  Please provide crash log, stack trace or a description of the error.  Also, please provide information about what platform and version of cordova you are using.

P.S. It sounds like you are adding this as a third-party plugin in the XDK?  You should be able to use r1.2.4 as a workaround.

**Comment by garakh** [comment]
@tony-- 
cordova: "3.6.3"
model: "ME173X"
platform: "Android"
version: "4.2.2"
crosswalk

We enable plugin «intel.xdk.device» in intel xdk, and calling «scanBarcode()».

Trying to figure out how to pick error info up.

**Comment by tony--** [comment]
If you have android tools installed, you should be able to run `adb logcat` and see the error message in the console when the app crashes.
In any case, this should be sufficient for me to reproduce the issue - thanks for providing this information!

**Comment by garakh** [comment]
We don't have them installed. 
Only Intel XDK. We develop the app using JS and then build it using remote Intel builder which gives us the apk file we upload to a mobile.

**Comment by tony--** [comment]
I'll look into this issue ASAP, but in the meantime, you may be able to workaround it by adding the plugin as a third-party plugin instead of checking the checkbox next to Intel XDK Device plugin.
The UI might be different depending on the version of the XDK you are using, but in XDK 2366 it looks like this:
![featured-plugins](https://cloud.githubusercontent.com/assets/513830/9959047/16056bec-5dde-11e5-8140-1a07f631424b.png)
![third-party-plugins](https://cloud.githubusercontent.com/assets/513830/9959046/16052d3a-5dde-11e5-9fd6-797ea5bebcda.png)
In third-party plugins section, click "Add plugin from web" and use the following settings:
Intel XDK Device
intel.xdk.device
https://github.com/01org/cordova-plugin-intel-xdk-device#r1.2.4
This should cause your app to get 1.2.4 (the previous version) instead of 1.2.5 (the version that is crashing for you).

**Comment by xmnboy** [comment]
The XDK includes a copy of adb, you may have to search your system to find it, but it is there. If you have used the Debug tab successfully your adb works. If you have not, follow the instructions to make the debug tab work and then you can use adb to perform the adb logcat operation Tony is requesting. 
On Sep 18, 2015 5:10 AM, Iliya Garakh notifications@github.com wrote:We don't have them installed. 
Only Intel XDK. We develop the app using JS and then build it using remote Intel builder which gives us the apk file we upload to a mobile.

**Comment by tony--** [comment]
I reproduced the issue locally via Corova CLI with cordova-android@3.6.3 (non-crosswalk) and Intel XDK Device plugin version 1.2.5.
@garakh - How were you including the plugin in your project in the XDK?  Were you just checking the checkbox next to the plugin?  Or were you adding it as a third-party plugin?  
Also, are you getting a build error or does your app crash when you run it?

**Comment by garakh** [comment]
@tony-- @xmnboy we have tried out all variants (built-in plugin, local and web link) and always got a crash. The app runs and works fine but it crashes when we try to run a camera using method scanBarcode()

**Comment by garakh** [comment]
We have checked app using  «intel app preview crosswalk» in debug mode and it works there. But it doesn't work when app is built to apk

**Comment by tony--** [comment]
@garakh Based on your previous comments, it seems like somehow you are getting the latest version of the plugin from the repo instead of a tagged version.  I'm not sure how your project is configured so that this happens and would like more information about your project configuration.
In any case, I pushed an update that bumps the version to 1.2.6-dev and introduces a compatibility fix that should prevent crash with cordova-android 3.6.3, cordova-android 4.1.0, crosswalk-cordova-android 3.6.3 and cordova-plugin-crosswalk-webview.  I hope that it will fix the issue you are seeing.
If it does not, then I think the problem you are having may be due to some other issue, possibly due to changes on the backend in our cloud builders.  
Please let me know what happens after you try another build.

**Comment by tony--** [comment]
I'd also like to suggest that you post a message on [the Intel XDK forums](https://software.intel.com/en-us/forums/intel-xdk) linking to this issue and mention or send a PM to [me](https://software.intel.com/en-us/user/1076877) and/or to [xmnboy](https://software.intel.com/en-us/user/436096).  If it turns out that the trouble you are having is not directly related to this plugin but is instead a problem with the Intel XDK build system, we will be able to help you more effectively there.

**Comment by xmnboy** [comment]
@garakh -- the Debug tab and App Preview have an older version of that plugin installed within them, thus the reason they work. The "local copy" approach that Tony mentions to include the plugin should work, I'm surprised it doesn't. 

The best way to control the version of the plugin used by your app is to use the EA version or to wait for our next update to the XDK, which will be very soon. Both use a much different technique to manage the plugin versions, by copying them directly into your project directory (using a behind the scenes "plugin add" to add them directly to your project directory). 

The quickest way to try this is with the EA release (see this forum post https://software.intel.com/en-us/forums/intel-xdk/topic/561064). Make a copy of your entire project directory and use "Open an Intel XDK Project" at the bottom left of the Projects tab (after opening the EA release) to open that copied project, some "upgrading" of the project configuration files will take place (your sources will not be changed).

Once open in EA, goto the "plugin management" screen via the Test tab and select "third-party plugins" (do not use the "custom intel xdk plugins" option). Once there, use:
- "plugin source" equals Git Repo
- "repo url" equals https://github.com/01org/cordova-plugin-intel-xdk-device.git
- "git ref" equals "r1.2.4" or any other earlier version [listed here](https://github.com/01org/cordova-plugin-intel-xdk-device/releases)

This will pull in the specific version you request, into the plugins directory of your EA project. That specific version will be used by the build system to build your app.

**Comment by garakh** [comment]
@tony-- @xmnboy 
We went deeper to XDK logs and found out:
1. Crosswalk with native intel.xdk.device crashes on scanBarcode(). 
Accordingly to the logs XDK uses **intel.xdk.device v1.0.7**. There is no such release at github.
New commit was pushed at the same time we began getting the crash that's why we decided that  commit had a trouble. But now we see that XDK uses v1.0.7 and we don't know if it was changed few days ago. 
We tried to use local and web plugins (1.2.6-dev from web, 1.2.5 and 1.2.4) and got the same error (the app crashes attempting to use barcode scanner).
But it makes sense now because XDK used to use 1.0.7 (we can't check 1.0.6 because it's not in git)
We also checked Cordova version — it's 3.6.3
1. We switched to Android without Crosswalk and everything works fine here, barscanner works properly (with native intel.xdk.device). Btw in this case XDK uses 1.0.8 release of intel.xdk.device.
   Cordova version due to setting page is 4.1.2 
   We tried to setuo 3.6.3 but got an error that 3.6.3 is not supported but XDK.
2. Sometimes local pluging was a reason of build crashes. But we couldnt stable reproduce this.
3. XDK EA doesn't work at all. Error window eternally comes up asking to rerun XDK.

Looks like it's not a problem of plugin, something has been changed in Remote Intel Builder which causes this bug.

p.s. why does XDK builder use such old plugin?
